include/electionguard/collections.h

```diff
@@ -0,0 +1,93 @@
+/// @file collections.h
+#ifndef __ELECTIONGUARD_CPP_COLLECTIONS_H_INCLUDED__
+#define __ELECTIONGUARD_CPP_COLLECTIONS_H_INCLUDED__
+
+#include "export.h"
+#include "status.h"
+
+#include <stdbool.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief A singly-linked list data structure composed of strings for keys and strings for values.
+ */
+typedef struct eg_linked_list_s eg_linked_list_t;
+
+/**
+* @brief Create a new linked list
+* @param[out] handle A pointer to the `eg_linked_list_t` opaque instance
+                     the caller is responsible for freeing the string.
+* @return eg_electionguard_status_t indicating success or failure
+* @retval ELECTIONGUARD_STATUS_SUCCESS The function was successfully executed
+* @retval ELECTIONGUARD_STATUS_ERROR_BAD_ALLOC The function was unable to allocate memory
+*/
+EG_API eg_electionguard_status_t eg_linked_list_new(eg_linked_list_t **out_handle);
+
+EG_API eg_electionguard_status_t eg_linked_list_free(eg_linked_list_t *handle);
+
+/**
+* @brief Append the key value pair to the end of the linked list
+* @param[in] handle A pointer to the `eg_linked_list_t` opaque instance
+* @param[in] in_key A string value to be used as a key
+* @param[in] in_value A string value to be used as a value
+* @return eg_electionguard_status_t indicating success or failure
+* @retval ELECTIONGUARD_STATUS_SUCCESS The function was successfully executed
+* @retval ELECTIONGUARD_STATUS_ERROR_BAD_ALLOC The function was unable to allocate memory
+*/
+EG_API eg_electionguard_status_t eg_linked_list_append(eg_linked_list_t *handle, char *in_key,
```

Comment:
at this point you will probably notice a few things about these function declarations:
- they are namespaced `eg_electionguard` or `eg_someclass_somemethod`
- the types are labelled as a type (e.g. the `_t` on the ends of types
- all of the function signatures follow the same pattern of (context, input_params, output_params)
- all of the input params start with `in_` and are always pointers for reference types
- all of the output params are called `out_` and the comment tells the caller if they are responsible for memory management.
- some output params also include `_ref` at the end to let the caller know that the object they have a handle to is a reference and the memory is not owned by them.

Reply:
you'll laso notice that we always return `eg_electionguard_status_t` for anything that can throw and for all reference type output values. in some circumstances a function may return a uint64_t or a bool directly if it does not throw an exception, or if its exceptional state is recoverable. this is all to make up for the fact that we can't handle exceptions across the binary interface on all platforms and C has no built in support for exceptions.

Reply:
^ is super helpful and would make an excellent `styleguide.md` or some such for guidance to others in the community that may work on this project, and also as something to point them to if a PR doesn't match (e.g. "Great PR bro, please just fix it to match the style guide").